middle_end/flambda2/from_lambda/closure_conversion.ml

```diff
@@ -471,17 +677,42 @@ let close_let acc env id user_visible defining_expr
     | Some (Simple simple) ->
       let body_env = Env.add_simple_to_substitute env id simple in
       body acc body_env
-    | Some _ | None -> (
+    | None ->
       (* CR pchambart: Not tail ! *)
       let acc, body = body acc body_env in
-      match defining_expr with
-      | None -> acc, body
-      | Some defining_expr ->
-        let var = VB.create var Name_mode.normal in
-        Let_with_acc.create acc
-          (Bound_pattern.singleton var)
-          defining_expr ~body
-        |> Expr_with_acc.create_let)
+      acc, body
+    | Some defining_expr ->
+      let body_env =
+        match defining_expr with
+        | Prim (Variadic (Make_block (_, Immutable), fields), _) ->
+          let approxs =
+            List.map (Env.find_value_approximation body_env) fields
+            |> Array.of_list
+          in
+          Env.add_block_approximation body_env (Name.var var) approxs
+        | Prim (Binary (Block_load _, block, field), _) ->
+          begin match Env.find_value_approximation body_env block with
+          | Value_unknown -> body_env
+          | Closure_approximation _ -> assert false
+          | Block_approximation approx ->
+            let approx : Env.value_approximation =
+              Simple.pattern_match field
+                ~const:(fun const ->
+                  match Reg_width_things.Const.descr const with
+                  | Tagged_immediate i ->
+                    approx.(Targetint_31_63.(Imm.to_int (to_targetint i)))
```

Comment:
I think you should handle the case where `i` is greater than `Array.length approx` explicitly, to get a better error message than an `Index_out_of_bounds` exception